java/ql/src/semmle/code/java/dataflow/internal/DataFlowImplCommon.qll

```diff
@@ -1,915 +1,866 @@
-          compatibleTypes(getErasedNodeType(p), getErasedNodeType(node)) and
-          not isUnreachableInCall(node, cc.(CallContextSpecificCall).getCall())
+        // flow through: possible prior read and prior store with compatible
+        // flow-through method
+        exists(ArgumentNode arg, ContentOption contentMid |
+          parameterValueFlowArg(p, arg, contentIn, contentMid) and
+          argumentValueFlowsThrough(arg, node, contentMid, contentOut)
         )
       }
 
-      /**
-       * Holds if `p` can flow to a return node of kind `kind` in the same
-       * callable using only value-preserving steps, in call context `cc`.
-       */
-      cached
-      predicate parameterValueFlowsThrough(ParameterNode p, ReturnKind kind, CallContextCall cc) {
-        parameterValueFlow(p, getAReturnNodeOfKind(kind), cc)
+      pragma[nomagic]
+      private predicate parameterValueFlowArg(
+        ParameterNode p, ArgumentNode arg, ContentOption contentIn, ContentOption contentOut
+      ) {
+        parameterValueFlow(p, arg, contentIn, contentOut) and
+        Cand::argumentValueFlowsThroughCand(arg, _, _, _)
+      }
+
+      pragma[nomagic]
+      predicate parameterValueFlowsToPostUpdate(
+        ParameterNode p, PostUpdateNode n, ContentOption contentIn, ContentOption contentOut
+      ) {
+        parameterValueFlow(p, n, contentIn, contentOut) and
+        contentOut.hasContent()
       }
 
       pragma[nomagic]
-      private predicate argumentValueFlowsThrough0(
-        DataFlowCall call, ArgumentNode arg, ReturnKind kind, CallContext cc
+      predicate argumentValueFlowsThrough0(
+        DataFlowCall call, ArgumentNode arg, ReturnKindExt kind, ContentOption contentIn,
+        ContentOption contentOut
       ) {
-        exists(ParameterNode param, CallContext innercc |
-          viableParamArg(call, param, arg, cc, innercc) and
-          parameterValueFlowsThrough(param, kind, innercc)
+        exists(ParameterNode param | viableParamArg(call, param, arg) |
+          parameterValueFlowReturn(param, _, kind, contentIn, contentOut)
         )
       }
 
       /**
        * Holds if `arg` flows to `out` through a call using only value-preserving steps,
-       * in call context cc.
+       * not taking call contexts into account.
+       *
+       * `contentIn` describes the content of `arg` that can flow to `out` (if any), and
+       * `contentOut` describes the content of `out` that it flows to (if any).
        */
-      cached
-      predicate argumentValueFlowsThrough(ArgumentNode arg, OutNode out, CallContext cc) {
-        exists(DataFlowCall call, ReturnKind kind |
-          argumentValueFlowsThrough0(call, arg, kind, cc)
-        |
-          out = getAnOutNode(call, kind) and
-          not isUnreachableInCall(out, cc.(CallContextSpecificCall).getCall()) and
-          compatibleTypes(getErasedNodeType(arg), getErasedNodeType(out))
+      private predicate argumentValueFlowsThrough(
```

Comment:
This predicate ought to be merged with the cached predicate of the same name, and the type check should occur here as part of the recursion.